File: src/unit/json2gpx.js
const moment = require('moment');
const xmlbuilder = require('xmlbuilder');

export function convertJsonToGpx(data) {
    const gpx = xmlbuilder.create('gpx', { version: '1.0', encoding: 'UTF-8' })
      .att('xmlns', 'http://www.topografix.com/GPX/1/1')
      .att('version', '1.1')
      .att('creator', 'Your Application');
  
    const gpxTrack = gpx.ele('trk');
    const gpxSegment = gpxTrack.ele('trkseg');
  
    data.forEach((item) => {
      const { latitude, longitude, altitude, heading, speed, timestamp } = item.coords;
      const timestampStr = item.timestamp;
  
      // Convert timestamp from string to Date object
      const parsedTimestamp = moment.utc(timestampStr, moment.ISO_8601).toDate();
  
      const gpxTrackpoint = gpxSegment.ele('trkpt', { lat: latitude, lon: longitude });
      gpxTrackpoint.ele('ele', {}, altitude);
      gpxTrackpoint.ele('time', {}, parsedTimestamp.toISOString());
      gpxTrackpoint.ele('course', {}, heading);
      gpxTrackpoint.ele('speed', {}, speed);
    });
  
    return gpx.end({ pretty: true });
  }

File: json2gpx.py
import sys
import json
from datetime import datetime
import gpxpy
import gpxpy.gpx

def convert_json_to_gpx(data):
    gpx = gpxpy.gpx.GPX()

    gpx_track = gpxpy.gpx.GPXTrack()
    gpx_segment = gpxpy.gpx.GPXTrackSegment()

    for item in data:
        latitude = item['coords']['latitude']
        longitude = item['coords']['longitude']
        ev=item['coords']['altitude']
        heading=item['coords']['heading']
        speed=item['coords']['speed']
        timestamp_str = item['timestamp']

        # Convert timestamp from string to datetime object
        timestamp = datetime.fromisoformat(timestamp_str.replace('Z', '+00:00'))

        gpx_trackpoint = gpxpy.gpx.GPXTrackPoint(latitude, longitude,elevation=ev, time=timestamp)
        gpx_trackpoint.course = heading
        gpx_trackpoint.speed = speed
        gpx_segment.points.append(gpx_trackpoint)
        
    gpx_track.segments.append(gpx_segment)
    gpx.tracks.append(gpx_track)

    return gpx.to_xml(version='1.0')

# Read JSON data from stdin
json_data = sys.stdin.read()

try:
    # Parse JSON
    data = json.loads(json_data)

    # Convert to GPX
    gpx_data = convert_json_to_gpx(data)

    # Output GPX data to stdout
    sys.stdout.write(gpx_data)
except json.JSONDecodeError:
    print("Invalid JSON format")
